File: app/src/main/java/com/lumi/dockeditor/AppListAdapter.java
package com.lumi.dockeditor;

import android.content.Context;
import android.content.pm.PackageManager;
import android.graphics.drawable.Drawable;
import android.view.LayoutInflater;
import android.view.MotionEvent;
import android.view.View;
import android.view.ViewGroup;
import android.widget.ImageView;
import android.widget.TextView;
import androidx.annotation.NonNull;
import androidx.recyclerview.widget.RecyclerView;
import java.util.List;

public class AppListAdapter extends RecyclerView.Adapter<AppListAdapter.AppViewHolder> {
    private List<AppInfo> apps;
    private OnStartDragListener dragListener;
    private OnItemClickListener clickListener;
    private OnItemRemoveListener removeListener;

    public interface OnStartDragListener {
        void onStartDrag(RecyclerView.ViewHolder viewHolder);
    }
    
    public interface OnItemClickListener {
        void onItemClick(int position);
    }
    
    public interface OnItemRemoveListener {
        void onItemRemove(int position);
    }
    
    public AppListAdapter(List<AppInfo> apps, OnStartDragListener dragListener,
                         OnItemClickListener clickListener, OnItemRemoveListener removeListener) {
        this.apps = apps;
        this.dragListener = dragListener;
        this.clickListener = clickListener;
        this.removeListener = removeListener;
    }
    
    @NonNull
    @Override
    public AppViewHolder onCreateViewHolder(@NonNull ViewGroup parent, int viewType) {
        View view = LayoutInflater.from(parent.getContext())
            .inflate(R.layout.item_app, parent, false);
        return new AppViewHolder(view);
    }
    
    @Override
    public void onBindViewHolder(@NonNull AppViewHolder holder, int position) {
        AppInfo app = apps.get(position);
        holder.bind(app);
    }
    
    @Override
    public int getItemCount() {
        return apps.size();
    }
    
    public class AppViewHolder extends RecyclerView.ViewHolder {
        private ImageView appIcon;
        private TextView appName;
        private TextView packageName;
        private TextView activityName;
        private ImageView dragHandle;
        private ImageView removeButton;
        
        public AppViewHolder(@NonNull View itemView) {
            super(itemView);
            appIcon = itemView.findViewById(R.id.appIcon);
            appName = itemView.findViewById(R.id.appName);
            packageName = itemView.findViewById(R.id.packageName);
            activityName = itemView.findViewById(R.id.activityName);
            dragHandle = itemView.findViewById(R.id.dragHandle);
            removeButton = itemView.findViewById(R.id.removeButton);
        }
        
        public void bind(AppInfo app) {
            appName.setText(app.getDisplayName());
            packageName.setText(app.packageName);
            activityName.setText(app.activity.isEmpty() ? "Default Activity" : app.activity);

            // **UPDATED ICON LOGIC**
            // Use the new, smarter method to get the display icon
            appIcon.setImageDrawable(app.getDisplayIcon(itemView.getContext()));
            
            // Click to edit app
            itemView.setOnClickListener(v -> {
                if (clickListener != null) {
                    clickListener.onItemClick(getAdapterPosition());
                }
            });
            
            // Remove button
            removeButton.setOnClickListener(v -> {
                if (removeListener != null) {
                    removeListener.onItemRemove(getAdapterPosition());
                }
            });
            
            // Drag handle for reordering
            dragHandle.setOnTouchListener((v, event) -> {
                if (event.getActionMasked() == MotionEvent.ACTION_DOWN) {
                    if (dragListener != null) {
                        dragListener.onStartDrag(this);
                    }
                }
                return false;
            });
        }
    }
}

File: app/src/main/java/com/lumi/dockeditor/AppListAdapter.kt
package com.lumi.dockeditor

import android.view.LayoutInflater
import android.view.MotionEvent
import android.view.View
import android.view.ViewGroup
import android.widget.ImageView
import android.widget.TextView
import androidx.recyclerview.widget.RecyclerView

class AppListAdapter(
    private val apps: List<AppInfo>,
    private val dragListener: OnStartDragListener?,
    private val clickListener: OnItemClickListener?,
    private val removeListener: OnItemRemoveListener?
) : RecyclerView.Adapter<AppListAdapter.AppViewHolder>() {

    /**
     * Interface for initiating drag-and-drop actions.
     */
    interface OnStartDragListener {
        fun onStartDrag(viewHolder: RecyclerView.ViewHolder)
    }

    /**
     * Interface for handling item clicks.
     */
    interface OnItemClickListener {
        fun onItemClick(position: Int)
    }

    /**
     * Interface for handling item removal.
     */
    interface OnItemRemoveListener {
        fun onItemRemove(position: Int)
    }

    override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): AppViewHolder {
        val view = LayoutInflater.from(parent.context)
            .inflate(R.layout.item_app, parent, false)
        return AppViewHolder(view)
    }

    override fun onBindViewHolder(holder: AppViewHolder, position: Int) {
        holder.bind(apps[position])
    }

    override fun getItemCount(): Int = apps.size

    /**
     * ViewHolder class that manages individual item views and their click/touch events.
     */
    inner class AppViewHolder(itemView: View) : RecyclerView.ViewHolder(itemView) {
        private val appIcon: ImageView = itemView.findViewById(R.id.appIcon)
        private val appName: TextView = itemView.findViewById(R.id.appName)
        private val packageNameText: TextView = itemView.findViewById(R.id.packageName)
        private val activityName: TextView = itemView.findViewById(R.id.activityName)
        private val dragHandle: View = itemView.findViewById(R.id.dragHandle)
        private val removeButton: View = itemView.findViewById(R.id.removeButton)

        fun bind(app: AppInfo) {
            // Setting text fields using property access syntax
            appName.text = app.getDisplayName()
            packageNameText.text = app.packageName
            activityName.text = if (app.activity.isEmpty()) "Default Activity" else app.activity

            // Load icon using the logic defined in AppInfo
            appIcon.setImageDrawable(app.getDisplayIcon(itemView.context))

            // Item click to edit the app
            itemView.setOnClickListener {
                val position = adapterPosition
                if (position != RecyclerView.NO_POSITION) {
                    clickListener?.onItemClick(position)
                }
            }

            // Click to remove app from the list
            removeButton.setOnClickListener {
                val position = adapterPosition
                if (position != RecyclerView.NO_POSITION) {
                    removeListener?.onItemRemove(position)
                }
            }

            // Handle touch events on the drag handle for reordering
            dragHandle.setOnTouchListener { _, event ->
                if (event.actionMasked == MotionEvent.ACTION_DOWN) {
                    dragListener?.onStartDrag(this)
                }
                false
            }
        }
    }
}